This project combines Generative AI and  Logic AI to form an integrated process, a pipeline, that retrieves large volumes of scientific information and produces logical models that can be validated and leveraged by human experts. Our goals are to organize and to analyze, with the assistance of AI and other bio-information retrieval tools, the documentation associated with a domain and extract validated knowledge to support planning for problem solving. We have tested our methodology and the pipeline with an experiment on COVID-19 which we are also prepared to report. The tool helped a team of biologists  to review the state of the art and handle the explosion of research papers during pandemia. The main product of the pipeline is a Gene Regulatory Network (GRN); a semantic map which constitutes a representation of causal relations in system biology which can in turn be used to explore different strategies for intervention and treatment. In a GRN, a regulatory event can activate a product that participates in an event, which activates or inhibits another product. Thus a GRN can be used to answer questions like: given a drug, what kinds of protein-protein interactions result when that drug binds to a known receptor, and which of them lead to the activation (or inhibition) of the transcriptional response of some gene, or the up/down regulation of a disease?. The pipeline (named Biopatternsg) has been structured around three main tasks: a) data and information gathering, from bioinformatics portals such as MeSH (ncbi.nlm.nih.gov/mesh), and GeneOntology (geneontology.org); b) the use of generative AI and Large Language Models (LLMs) for the automatic extraction of biological entities and their biological relationships, based on the services of PubTator (ww.ncbi.nlm.nih.gov/research/pubtator3); and c) the use of logic AI to manage the information obtained in a) and b) and combine it with conserved criteria, all represented as knowledge bases in Prolog (http://www.swi-prolog.org) and constraint logic programming. The system also keeps track of and allows access to all the original documents used as sources of knowledge. It is possible, for instance, to choose an inferred regulatory pathway and retrieve the abstracts that support it. A biological analysis conducted by human experts has concluded that the pipeline is useful for deepening the understanding of COVID-19. They also suggest the need to model the task of relation extraction at the sentence level, as opposed to the current modeling approach, developed at the abstract level. We are working to improve the system as suggested and further validate its effectiveness.

Biopatternsg is a system that builds gene regulatory networks from a basic collection of biological objects and DNA data, gathering and processing PubMed abstracts, using Internet available informational resources like Mesh, GeneOntology, UniProt, PDB and others, and that uses logical engines, prolog and java code, to integrate and analyze them. To carry out the modeling of such networks, the scientific community uses and develops various computer services available through portals available on the Internet. Some of these portals are: GeneOntology, PDB, HGNC, Pathway Commons, UniProt, PubMed, among others. The mentioned sites provide services that make possible to access such services automatically and therefore, these can be used to organize knowledge bases that integrate their resources. Our team has developed a system that allows such integration and the analysis of the information obtained in different modalities. Services such as automatic identity modeling (e.g. receptor, enzyme, etc.), molecular functions, and biological processes for the objects in a network and their protein-protein interactions, have been implemented. Our work describes an ontological and logical alternative for the discovery of biological signaling pathways and regulatory subnetworks within an GRN. This document describes the system’s capabilities and modes of use, which we hope will facilitate the definition of adjustments and new requirements, aimed at improving the utility of the system we have called biopatternsg (biopatterns searching). For this purpose, an example of modeling and analysis is used, in which it is desired to explore possible links between the regulatory processes inherent in SARS-COV and SARS-COV-2. At the moment the example mentioned only illustrates the use of the system, therefore its results are for academic purposes only. This document includes an annex that allows observing the trace of the bioPatternsg execution and the knowledge bases it generates, as well as details regarding its downloading and installation.

Please, do not hesitate to contact us at this email address: biopatternsg@gmail.com, for any problem,comment or doubt.

How to Cite: Lopez J., Ramirez Y., Dávila J., Bastidas M. A logical and ontological framework for knowledge discovery on gene regulatory networks. Case study: Bile Acid and Xenobiotic System (BAXS). Journal of Bioinformatics and Genomics, [S.l.], n. 2 (14), dec. 2020. ISSN 2530-1381. Available at: doi: http://dx.doi.org/10.18454/jbg.2020.2.14.1. Date accessed: 09 May. 2025.

Please, see the wiki for details about the system's installation and the system's user's guide.
